hazelcast/src/main/java/com/hazelcast/json/internal/JsonSchemaHelper.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2008-2019, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.json.internal;
+
+import com.fasterxml.jackson.core.JsonLocation;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import com.hazelcast.core.HazelcastException;
+import com.hazelcast.internal.json.JsonReducedValueParser;
+import com.hazelcast.internal.json.JsonValue;
+import com.hazelcast.internal.json.ParseException;
+import com.hazelcast.internal.serialization.impl.NavigableJsonInputAdapter;
+import com.hazelcast.query.impl.getters.JsonPathCursor;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.Reader;
+
+/**
+ * Responsible for creating and using {@link JsonSchemaNode}s.
+ */
+public final class JsonSchemaHelper {
+
+    private JsonSchemaHelper() {
+        // private constructor
+    }
+
+    /**
+     * Creates a {@link JsonPattern} for given json value. If given path
+     * has "any", pattern is created only upto that part. The rest is
+     * omitted.
+     *
+     * @param input         a buffer containing the queried object.
+     * @param schemaNode the description of the given object
+     * @param path          query path
+     * @return              a list of integers that represent the order
+     *                      of each query term within the queried object
+     * @throws IOException
+     */
+    @SuppressWarnings("checkstyle:npathcomplexity")
+    public static JsonPattern createPattern(NavigableJsonInputAdapter input, JsonSchemaNode schemaNode, JsonPathCursor path) {
```

Comment:
As far as I understand, the schema parameter for this method and `findValueWithPattern` needs to always match the input parameter, right? But the query path is allowed to differ (e.g. point to a non-terminal node, point to a node past a terminal node, not match the schema, ...)?

Reply:
Yes the schema always needs to match the input. We create the schemas from the input at put time. However, path may be different as it is created from the user provided string.

This method tries to create a valid pattern for given path only when the queried path is valid.

findValueWithPattern also accepts user provided path. In that case, we have seen a pattern with the same path before on another json object. However, the same pattern may not hold for the current object. So we check the integrity of our pattern with user provided path as we move along the object. If the pattern does not match user provided path there, that pattern has no use for that json object.